### DISCUSSION TITLE
Simplemonitor Suitable for my usecase?

### DISCUSSION THREAD

**Discussion by KerstinKeller** [jamesoff/simplemonitor#663] [Q&A]
Hi,

we're considering using Simplemonitor for a usecase of a distributed System where every system does their own bit of monitoring (e.g. monitoring connected hardware, monitoring running applications).
One system should collect all others systems monitoring, and take care of displaying / alterting.
I have run Simplemonitor with a simple configuration, and was able to get it up and running easily.

As some of the monitoring might be special, we will certainly have to write our own monitors. Is it easy to add other monitors? Is there some information about it? This point would be essential for us to be able to use Simplemonitor.

Do you think we can achieve our goal with the help of Simplemonitor, or should we look elsewhere?
Thank you!

**Comment by jamesoff** [comment]
Hi :)

Your use case is exactly the sort of thing I started SimpleMonitor for originally. Glad you've been able to get it running ok.

Writing your own monitors is pretty straightforward; you'll need some Python knowledge but it's not too involved (depending on how complex the actual monitoring is, I suppose). For slightly simpler things you can just write a shell script and use the "Command" monitor to execute it and validate the output or return code.

At the moment I haven't written any docs on creating your own Monitors, but I should be able to put something together pretty quickly if it would help.

**Reply by KerstinKeller** [reply]
Hi,

thanks for the quick reply. I think from the files that I had a look at, that I can stitch something together, without additional information. But if you do have time to write something, it will certainly be helpful.

One thing that I am wondering though, if the plugins will be detected if the Monitor was installed through pip. (Method which I followed). I guess they will only be discovered, if they are located into the `monitors` subfolder? (I haven't verified this, but I assume).

**Comment by jamesoff** [comment]
Yeah I'd suggest working from a cloned git repo for development. For best results, install `poetry` with pip, and then clone the repo, switch to the `develop` branch, and in that directory do `poetry install`. It'll be installed such that you can edit the files in the directory and those changes are immediately reflected when you run it. You can do `poetry build` to get a pip packge you can install on your other hosts to test (saved in `dist/`).

I think that should be right; I recently switched to poetry but I just validated that sequence seemed to work for me on a fresh clone! :)

See https://github.com/jamesoff/simplemonitor/commit/984d2c5404b2fad2aa6c170cf96be88f820452db for an example commit which adds a Monitor; broadly, you need to subclass the `Monitor` class, provide `run_test()` and `describe()` methods in it. The whole class needs to be decorated with `@register`, and add it as an import to `Monitors/__init__.py`. It'll then be discovered and usable. Alerters and Loggers are similar to add, if you need.

Hope that all makes sense, let me know if not. If your changes are suitable to share, I'd love a PR :)

**Reply by jamesoff** [reply]
Oh I think I messed up replying as a thread :(
